File: src/main/java/org/onlineoptimisation/algorithm/RandomSearch.java
package org.onlineoptimisation.algorithm;

import java.util.ArrayList;
import java.util.List;

import org.armadillojava.Arma;
import org.armadillojava.Col;
import org.armadillojava.Op;

public class RandomSearch extends OptimisationAlgorithm {

  @Override
  protected List<Col> getNextParameters() {
    List<Col> nextParameters = new ArrayList<>();

    Col parameter = Arma.randu(Col.class, _optimisationProblem.getNumberOfDimensions());
    parameter.inPlace(Op.ELEMTIMES, _optimisationProblem.getUpperBounds().minus(_optimisationProblem.getLowerBounds()));
    parameter.inPlace(Op.PLUS, _optimisationProblem.getLowerBounds());
    
    nextParameters.add(parameter);
    return nextParameters;
  }
}

File: src/optimisationAlgorithm/samplingBasedAlgorithm/randomSearch.cpp
#include <hop_bits/optimisationAlgorithm/samplingBasedAlgorithm/randomSearch.hpp>

namespace hop {
  RandomSearch::RandomSearch(
      const std::shared_ptr<OptimisationProblem> optimisationProblem) noexcept
    : SamplingBasedAlgorithm(optimisationProblem) {

  }

  void RandomSearch::optimiseImplementation() noexcept {
    bestSoftConstraintsValue_ = std::numeric_limits<double>::infinity();
    bestObjectiveValue_ = std::numeric_limits<double>::infinity();

    do {
      ++numberOfIterations_;

      const arma::Col<double>& candidateParameter = arma::randu<arma::Col<double>>(optimisationProblem_->getNumberOfDimensions()) % (optimisationProblem_->getUpperBounds() - optimisationProblem_->getLowerBounds()) + optimisationProblem_->getLowerBounds();
      const double& candidateSoftConstraintsValue = optimisationProblem_->getSoftConstraintsValue(candidateParameter);
      const double& candidateObjectiveValue = optimisationProblem_->getObjectiveValue(candidateParameter);

      if(candidateSoftConstraintsValue < bestSoftConstraintsValue_ || candidateSoftConstraintsValue == bestSoftConstraintsValue_ && candidateObjectiveValue < bestObjectiveValue_) {
        bestParameter_ = candidateParameter;
        bestSoftConstraintsValue_ = candidateSoftConstraintsValue;
        bestObjectiveValue_ = candidateObjectiveValue;
      }
    } while(!isFinished() && !isTerminated());
  }

  std::string RandomSearch::to_string() const noexcept {
    return "RandomSearch";
  }
}
